File: kotlin/triangle/src/main/kotlin/Triangle.kt
class Triangle<out T : Number>(val a: T, val b: T, val c: T) {
  init {
    val IsTriangle: Boolean = (a.toDouble() + b.toDouble() > c.toDouble()) &&
                              (a.toDouble() + c.toDouble() > b.toDouble()) &&
                              (b.toDouble() + c.toDouble() > a.toDouble())
    if (!IsTriangle) {
      throw IllegalArgumentException("It is not a triangle")
    }
  }
   val isEquilateral: Boolean = (a == b && a == c && b == c) 
  val isIsosceles: Boolean = (a == b && a != c) ||
                             (a == c && a != b) ||
                             (b == c && a != b) ||
                             isEquilateral
  val isScalene: Boolean = (a != b && a != c && b != c)
}

File: prolog/triangle/triangle.pl
triangle(Side1, Side2, Side3, Type) :-
    is_triangle(Side1, Side2, Side3), is_right_type(Side1, Side2, Side3, Type).
    

is_triangle(Side1, Side2, Side3):- 
    (Side1 =< 0 ->
        fail
    ; Side2 =< 0 ->
        fail
    ; Side3 =< 0 ->
        fail
    ;
    Side1 + Side2 >= Side3, Side1 + Side3 >= Side2, Side2 + Side3 >= Side1
).

is_right_type(Side1, Side2, Side3, Type):-
    ( Type == "equileteral" ->
        is_equileteral(Side1, Side2, Side3)
      ; Type == "isosceles" ->
        is_isosceles(Side1, Side2, Side3)
      ; Type == "scalene" ->
        is_scalene(Side1, Side2, Side3)
    ).

is_equileteral(Side1, Side2, Side3).

is_isosceles(Side1, Side2, Side3):-
    Side1 =:= Side2; Side2 =:= Side3.

is_scalene(Side1, Side2, Side3):-
    Side1 \= Side2, Side1 \= Side2, Side3 \= Side2.

and(A,B):- is_true(A), is_true(B).